internal/test_unit/python/unit.py

```diff
@@ -0,0 +1,371 @@
+#!/usr/bin/python
+
+from datetime import datetime as dt
+import logging
+import os
+from pathlib import Path
+import re
+import subprocess
+import sys
+import xml.etree.ElementTree as ET
+
+def unit(test_xml, file_log=None, cmd_only=False, noexit=False, memchk=False, callchk=False):
+    """
+    Parse a unit test xml file, run the associated tests, and display test results.
+
+    Parameters
+    -----------
+    test_xml : pathlike
+        path to file containing the unit test(s) to perform
+    file_log : pathlike, default None
+        if present, write output from each test to the specified file
+    cmd_only : bool, default False
+        if true, print the test commands but do not run them (overrides file_log)
+    noexit : bool, default False
+        if true, the unit tester will continue executing subsequent
+                tests when a test fails
+    memchk : bool, default False
+        if true, activate valgrind with memcheck
+    callchk : bool, default False
+        if true, activate valgrind with callcheck
+    """
+  
+    # initialize logger
+    logger = logging.getLogger(__name__)
+    logger.setLevel(logging.DEBUG)
+
+    # create/add console handler
+    ch = logging.StreamHandler()
+    ch.setLevel(logging.INFO)
+    logger.addHandler(ch)
+
+    # create/add file handler
+    if file_log and not cmd_only:
+        fh = logging.FileHandler(file_log)
+        fh.setLevel(logging.DEBUG)
+        logger.addHandler(fh)
+   
+    # parse xml file
+    try:
+        test_root = ET.parse(test_xml)
+    except Exception as e:
+        logger.exception(f"ERROR: Unable to parse xml from {test_xml}")
+        raise
+
+    # parse the children of the met_test element
+    if test_root.getroot().tag != 'met_test':
+        logger.error(f"ERROR: unexpected top-level element. Expected 'met_test', got '{test_root.tag}'")
+        sys.exit(1)
+    else:
+        # read test_dir
```

Comment:
The else here is not needed since you exit inside the if block. You can remove the else statement and move lines 60-69 left 4 spaces.

Reply:
good catch. done.